File: apps/api/app/schemas/punch_list.py
"""Pydantic schemas for Punch List Items."""
from datetime import date, datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class PunchListItemCreate(BaseModel):
    title: str
    description: Optional[str] = None
    location: Optional[str] = None
    category: str = "DEFICIENCY"
    priority: str = "NORMAL"
    assigned_to_sub_id: Optional[UUID] = None
    assigned_to_user_id: Optional[UUID] = None
    due_date: Optional[date] = None
    cost_code_id: Optional[UUID] = None
    drawing_reference: Optional[str] = None


class PunchListItemUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    location: Optional[str] = None
    category: Optional[str] = None
    priority: Optional[str] = None
    assigned_to_sub_id: Optional[UUID] = None
    assigned_to_user_id: Optional[UUID] = None
    due_date: Optional[date] = None
    cost_code_id: Optional[UUID] = None
    drawing_reference: Optional[str] = None


class PunchListCompleteRequest(BaseModel):
    completion_notes: Optional[str] = None


class PunchListVerifyRequest(BaseModel):
    verified: bool
    verification_notes: Optional[str] = None


class PunchListItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    number: int
    formatted_number: str
    title: str
    description: Optional[str] = None
    location: Optional[str] = None
    category: str
    priority: str
    status: str
    assigned_to_sub_id: Optional[UUID] = None
    assigned_to_sub_name: Optional[str] = None
    assigned_to_user_id: Optional[UUID] = None
    assigned_to_user_name: Optional[str] = None
    due_date: Optional[date] = None
    cost_code_id: Optional[UUID] = None
    drawing_reference: Optional[str] = None
    before_photo_ids: list = []
    after_photo_ids: list = []
    verification_photo_ids: list = []
    completion_notes: Optional[str] = None
    completed_by: Optional[UUID] = None
    completed_at: Optional[datetime] = None
    verification_notes: Optional[str] = None
    verified_by: Optional[UUID] = None
    verified_at: Optional[datetime] = None
    comments_count: int = 0
    created_by: Optional[UUID] = None
    created_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class PunchListListResponse(BaseModel):
    data: list[PunchListItemResponse]
    meta: dict

File: apps/web/src/types/punch-list.ts
export interface PunchListItem {
  id: string;
  project_id: string;
  number: number;
  formatted_number: string;
  title: string;
  description?: string | null;
  location?: string | null;
  category: string;
  priority: string;
  status: string;
  assigned_to_sub_id?: string | null;
  assigned_to_sub_name?: string | null;
  assigned_to_user_id?: string | null;
  assigned_to_user_name?: string | null;
  due_date?: string | null;
  cost_code_id?: string | null;
  drawing_reference?: string | null;
  before_photo_ids: string[];
  after_photo_ids: string[];
  verification_photo_ids: string[];
  completion_notes?: string | null;
  completed_by?: string | null;
  completed_at?: string | null;
  verification_notes?: string | null;
  verified_by?: string | null;
  verified_at?: string | null;
  comments_count: number;
  created_by?: string | null;
  created_by_name?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PunchListItemCreateInput {
  title: string;
  description?: string;
  location?: string;
  category?: string;
  priority?: string;
  assigned_to_sub_id?: string;
  assigned_to_user_id?: string;
  due_date?: string;
  drawing_reference?: string;
}
